diff --git a/lib/stream.js b/lib/stream.js
index edc3018332d999139e6b6573230e8c0553650bc0..cf29a5f5f6bfb3e6fb3c3351290e13794242871f 100644
--- a/lib/stream.js
+++ b/lib/stream.js
@@ -37,8 +37,130 @@ function getNodejsStreamFuncs() {
  * The stream functions that work in Cloudflare Workers
  */
 function getCloudflareStreamFuncs() {
+  const { EventEmitter } = require('events')
+
+  class CloudflareSocket extends EventEmitter {
+    constructor(ssl) {
+      super()
+      this.ssl = ssl
+      this.writable = false
+      this.destroyed = false
+      this._upgrading = false
+      this._upgraded = false
+      this._cfSocket = null
+      this._cfWriter = null
+      this._cfReader = null
+    }
+
+    setNoDelay() {
+      return this
+    }
+
+    setKeepAlive() {
+      return this
+    }
+
+    ref() {
+      return this
+    }
+
+    unref() {
+      return this
+    }
+
+    async connect(port, host, connectListener) {
+      try {
+        if (connectListener) this.once('connect', connectListener)
+        const options = this.ssl ? { secureTransport: 'starttls' } : {}
+        const mod = await import('cloudflare:sockets')
+        const connect = mod.connect
+        this._cfSocket = connect(`${host}:${port}`, options)
+        this._cfWriter = this._cfSocket.writable.getWriter()
+        this._addClosedHandler()
+        this._cfReader = this._cfSocket.readable.getReader()
+
+        if (this.ssl) {
+          this._listenOnce().catch((e) => this.emit('error', e))
+        } else {
+          this._listen().catch((e) => this.emit('error', e))
+        }
+
+        await this._cfWriter.ready
+        this.writable = true
+        this.emit('connect')
+        return this
+      } catch (e) {
+        this.emit('error', e)
+      }
+    }
+
+    async _listen() {
+      while (true) {
+        const { done, value } = await this._cfReader.read()
+        if (done) break
+        this.emit('data', Buffer.from(value))
+      }
+    }
+
+    async _listenOnce() {
+      const { done, value } = await this._cfReader.read()
+      if (!done) this.emit('data', Buffer.from(value))
+    }
+
+    write(data, encoding = 'utf8', callback = () => {}) {
+      if (data.length === 0) return callback()
+      if (typeof data === 'string') data = Buffer.from(data, encoding)
+      this._cfWriter.write(data).then(
+        () => callback(),
+        (err) => callback(err)
+      )
+      return true
+    }
+
+    end(data = Buffer.alloc(0), encoding = 'utf8', callback = () => {}) {
+      this.write(data, encoding, (err) => {
+        this._cfSocket.close()
+        if (callback) callback(err)
+      })
+      return this
+    }
+
+    destroy() {
+      this.destroyed = true
+      return this.end()
+    }
+
+    startTls(options) {
+      if (this._upgraded) {
+        this.emit('error', 'Cannot call `startTls()` more than once on a socket')
+        return
+      }
+      this._cfWriter.releaseLock()
+      this._cfReader.releaseLock()
+      this._upgrading = true
+      this._cfSocket = this._cfSocket.startTls(options)
+      this._cfWriter = this._cfSocket.writable.getWriter()
+      this._cfReader = this._cfSocket.readable.getReader()
+      this._addClosedHandler()
+      this._listen().catch((e) => this.emit('error', e))
+    }
+
+    _addClosedHandler() {
+      this._cfSocket.closed
+        .then(() => {
+          if (!this._upgrading) {
+            this._cfSocket = null
+            this.emit('close')
+          } else {
+            this._upgrading = false
+            this._upgraded = true
+          }
+        })
+        .catch((e) => this.emit('error', e))
+    }
+  }
+
   function getStream(ssl) {
-    const { CloudflareSocket } = require('pg-cloudflare')
     return new CloudflareSocket(ssl)
   }
 
